### PR TITLE
feat(expr): switch to `fancy-regex` crate & update the original version

### LINKED ISSUE #12119: expr: switch to fancy-regex crate
https://github.com/fancy-regex/fancy-regex

This crate supports more advanced features and is closer to PostgreSQL regex.

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #12119

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

The `fancy-regex` adds support for advanced regex features like back-reference & positive/negative lookahead, may need to update the documentation according to this.

</details>

### CODE REVIEW

**Comment by xzhseh** [comment]
The expression (regex) in `fancy-regex` will be essentially parsed to a `ExprTree` when building by `RegexBuilder`, thus we do not need to explicitly specify `-i` flag, but to integrate it with the input `pattern`.
Also, we support back-reference and positive/negative lookahead in the `regex` now, which means the following query should be valid at present: 
```sql
query T
select regexp_replace('foobarbaz', 'a(?=r)', 'X');
----
foobXrbaz
```

**Comment by xzhseh** [comment]
Part of `Regex::new_options` is as below:
```rust
fn new_options(options: RegexOptions) -> Result<Regex> {
        let raw_tree = Expr::parse_tree(&options.pattern)?;

        // wrapper to search for re at arbitrary start position,
        // and to capture the match bounds
        let tree = ExprTree {
            expr: Expr::Concat(vec![
                Expr::Repeat {
                    child: Box::new(Expr::Any { newline: true }),
                    lo: 0,
                    hi: usize::MAX,
                    greedy: false,
                },
                Expr::Group(Box::new(raw_tree.expr)),
            ]),
            ..raw_tree
        };
...
}
```

File: src/expr/src/vector_op/regexp.rs [hunk 1]
```diff
@@ -142,7 +152,7 @@ fn regexp_match(text: &str, regex: &RegexpContext) -> Option<ListValue> {
     // If there are multiple captures, then the first one is the whole match, and should be
     // ignored in PostgreSQL's behavior.
     let skip_first = regex.regex.captures_len() > 1;
-    let capture = regex.regex.captures(text)?;
+    let capture = regex.regex.captures(text).unwrap()?;
```
Comment (wangrunji0408) [inline src/expr/src/vector_op/regexp.rs:155]: In which case does the `captures()` return an error?

File: src/expr/src/vector_op/regexp.rs [hunk 2]
```diff
@@ -32,10 +32,20 @@ pub struct RegexpContext {
...
+        // FIXME: Any better solution?
+        let mut origin = String::with_capacity(pattern.len());
+
+        if options.case_insensitive {
+            origin.push_str("(?i:");
+            origin.push_str(pattern);
+            origin.push_str(")");
+        } else {
+            origin = pattern.to_string();
+        };
```
Comment (wangrunji0408) [inline src/expr/src/vector_op/regexp.rs:45]:
```suggestion
        let origin = if options.case_insensitive {
            format!("(?i:{})", pattern)
        } else {
            pattern.to_string()
        };
```

File: src/expr/src/vector_op/regexp.rs [hunk 3]
```diff
@@ -406,12 +416,12 @@ fn regexp_replace(
                 }
             } else {
                 // `N` is not specified
-                if ctx.regex.captures(&text[start..]).is_none() {
+                if ctx.regex.captures(&text[start..]).is_err() {
```
Comment (wangrunji0408) [inline src/expr/src/vector_op/regexp.rs:419]: Shouldn't it be like this?
```suggestion
                if let Ok(None) = ctx.regex.captures(&text[start..]) {
```

File: src/expr/src/vector_op/regexp.rs (hunk 1 above)
Comment (xzhseh) [reply src/expr/src/vector_op/regexp.rs:150]: > Returns the capture groups for the first match in text. If no match is found, then Ok(None) is returned.

Thus I think it's safe to call `unwrap()` here 🤔️?

Comment (wangrunji0408) [reply src/expr/src/vector_op/regexp.rs:150]: > Returns the capture groups for the first match in text. If no match is found, then Ok(None) is returned.

But the document above doesn't mention the error. 🤔️

**Comment by TennyZhuang** [comment]
Are there some inputs can make the expression panic?

**Review by TennyZhuang — APPROVED** [review]
Test cases LGTM

**Review by wangrunji0408 — APPROVED** [review]
LGTM

File: src/expr/src/vector_op/regexp.rs (hunk 1 above)
Comment (xzhseh) [reply src/expr/src/vector_op/regexp.rs:150]: Seems that the only `Err` will be generated from `captures` here is in the inner `fancy_regex::run`, which is a runtime error.
CleanShot 2023-09-15 at 13 16 35@2x

Comment (xzhseh) [reply src/expr/src/vector_op/regexp.rs:150]: And according to this `Err`, I prefer to let it panic since some really strange errors are happening 🤔️?

**Comment by xzhseh** [comment]
> Are there some inputs can make the expression panic?

Yes, like if the input pattern can not be parsed when trying to `RegexBuilder::build()`, an `Error` will be returned.
And in the current implementation we just simply call `unwrap()` on it, which will cause panic in this case.
We could definitely make a wrapper to handle this if necessary.

The below is all the `Error` that may be triggered during the backtracking regex approach.

```rust
/// An error as the result of parsing, compiling or running a regex.
#[derive(Debug)]
pub enum Error {
    /// An error as a result of parsing a regex pattern, with the position where the error occurred
    ParseError(ParseErrorPosition, ParseError),
    /// An error as a result of compiling a regex
    CompileError(CompileError),
    /// An error as a result of running a regex
    RuntimeError(RuntimeError),

    /// This enum may grow additional variants, so this makes sure clients don't count on exhaustive
    /// matching. Otherwise, adding a new variant could break existing code.
    #[doc(hidden)]
    __Nonexhaustive,
}

```

File: src/expr/src/vector_op/regexp.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/expr/src/vector_op/regexp.rs:150]: Okay!

**Pull request merged**